packages/flutter/lib/src/material/input_decorator.dart

```diff
@@ -804,109 +817,159 @@ class _RenderDecoration extends RenderBox {
+    // Margin on each side of subtext (counter and helperError)
+    final Map<RenderBox, double> boxToBaseline = <RenderBox, double>{};
+    final BoxConstraints boxConstraints = layoutConstraints.loosen();
+
+    // Layout all the widgets used by InputDecorator
+    boxToBaseline[prefix] = _layoutLineBox(prefix, boxConstraints);
+    boxToBaseline[suffix] = _layoutLineBox(suffix, boxConstraints);
+    boxToBaseline[icon] = _layoutLineBox(icon, boxConstraints);
+    boxToBaseline[prefixIcon] = _layoutLineBox(prefixIcon, boxConstraints);
+    boxToBaseline[suffixIcon] = _layoutLineBox(suffixIcon, boxConstraints);
+    // TODO(justinmc): In the old layout, there was inputWidth that constrained
+    // some of these, but including it here breaks tests for me.
+    boxToBaseline[label] = _layoutLineBox(label, boxConstraints);
+    boxToBaseline[hint] = _layoutLineBox(hint, boxConstraints);
+    boxToBaseline[counter] = _layoutLineBox(counter, boxConstraints);
+
+    // The helper or error text can occupy the full width less the space
+    // occupied by the icon and counter.
+    boxToBaseline[helperError] = _layoutLineBox(
+      helperError,
+      boxConstraints.copyWith(
         maxWidth: math.max(0.0, boxConstraints.maxWidth
           - _boxSize(icon).width
           - _boxSize(counter).width
           - contentPadding.horizontal,
         ),
-      );
-      layoutLineBox(helperError);
+      ),
+    );
 
-      if (aboveBaseline + belowBaseline > 0.0) {
-        const double subtextGap = 8.0;
-        subtextBaseline = containerHeight + subtextGap + aboveBaseline;
-        subtextHeight = subtextGap + aboveBaseline + belowBaseline;
-      }
+    // The height of the input needs to accommodate label above and counter and
+    // helperError below, when they exist.
+    const double SUBTEXT_GAP = 8.0;
+    final double labelHeight = label == null
+      ? 0
+      : decoration.floatingLabelHeight;
+    final double topHeight = decoration.border.isOutline
+        ? math.max(labelHeight - boxToBaseline[label], 0)
+        : labelHeight;
+    final double counterHeight = counter == null
+      ? 0 : boxToBaseline[counter] + SUBTEXT_GAP * 2;
+    final _HelperError helperErrorWidget = decoration.helperError;
+    final double helperErrorHeight = helperErrorWidget.helperText == null
+      ? 0 : helperError.size.height + SUBTEXT_GAP * 2;
+    final double bottomHeight = math.max(
+      counterHeight,
+      helperErrorHeight,
+    );
+    boxToBaseline[input] = _layoutLineBox(
+      input,
+      boxConstraints.deflate(EdgeInsets.only(
+        top: contentPadding.top + topHeight,
+        bottom: contentPadding.bottom + bottomHeight,
+      )),
+    );
+
+    // The field can be occupied by a hint or by the input itself
+    final double hintHeight = hint == null ? 0 : hint.size.height;
+    final double inputDirectHeight = input?.size?.height == null ? 0 : input.size.height;
```

Comment:
At this point input should either be null or have a size because it's been laid out. So: `input == null ? 0 : input.size.height`